File: compress.m
//
//  compress.c
//  buildo
//
//  Created by Michael Eisel on 10/20/21.
//

#include <compression.h>
#include <Foundation/Foundation.h>

int main(int argc, char *argv[]) {
    if (argc != 3) {
        return 1;
    }
    NSString *inFile = @(argv[1]);
    NSString *outFile = @(argv[2]);
    NSData *data = [NSData dataWithContentsOfFile:inFile options:NSDataReadingMappedIfSafe error:nil];
    if (!data) {
        return 1;
    }
    size_t outBufferLength = data.length + sizeof(uint64_t) + 500 * 1024;
    uint8_t *outBuffer = (uint8_t *)malloc(outBufferLength);
    uint64_t outLength = compression_encode_buffer(outBuffer + sizeof(uint64_t), outBufferLength - sizeof(uint64_t), data.bytes, data.length, NULL, COMPRESSION_LZFSE);
    if (outLength == 0) {
        NSLog(@"Error occurred: either during compression, or because resulting file is much larger than original file");
        return 1;
    }
    uint64_t inLength = data.length;
    memcpy(outBuffer, &inLength, sizeof(uint64_t));
    NSData *outData = [NSData dataWithBytesNoCopy:outBuffer length:outLength + sizeof(uint64_t) freeWhenDone:YES];
    [outData writeToFile:outFile atomically:YES];
    return 0;
}

File: compress.swift
guard CommandLine.argc == 3 else {
    exit(1)
}

let inFile = CommandLine.arguments[1] as NSString
let outFile = CommandLine.arguments[2] as NSString
let data = try? NSData(contentsOfFile: inFile as String, options: .mappedIfSafe)
guard let data = data else {
    exit(1)
}

let outBufferLength = data.length + MemoryLayout<Int>.stride + 500 * 1024
let outBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: outBufferLength)
let outLength = compression_encode_buffer(
    outBuffer.advanced(by: MemoryLayout<Int>.stride), outBufferLength - MemoryLayout<Int>.stride,
    data.bytes.assumingMemoryBound(to: UInt8.self), data.length, nil, COMPRESSION_LZFSE)
guard outLength > 0 else {
    NSLog("Error occurred: either during compression, or because resulting file is much larger than original file")
    exit(1)
}

var inLength = data.length
memcpy(outBuffer, &inLength, MemoryLayout<Int>.stride)
let outData = NSData(bytesNoCopy: outBuffer, length: outLength + MemoryLayout<Int>.stride, freeWhenDone: true)
try! outData.write(toFile: outFile as String, options: .atomic)
exit(0)
